sweep/fee_bumper.go

```diff
@@ -921,3 +928,235 @@ func calcCurrentConfTarget(currentHeight, deadline int32) uint32 {
+			continue
+		}
+
+		idxs = append(idxs, o)
+		sweepTx.AddTxIn(&wire.TxIn{
+			PreviousOutPoint: *o.OutPoint(),
+			Sequence:         o.BlocksToMaturity(),
+		})
+		sweepTx.AddTxOut(o.RequiredTxOut())
+	}
+
+	// Sum up the value contained in the remaining inputs, and add them to
+	// the sweep transaction.
+	for _, o := range inputs {
+		if o.RequiredTxOut() != nil {
+			continue
+		}
+
+		idxs = append(idxs, o)
+		sweepTx.AddTxIn(&wire.TxIn{
+			PreviousOutPoint: *o.OutPoint(),
+			Sequence:         o.BlocksToMaturity(),
+		})
+	}
+
+	// If there's a change amount, add it to the transaction.
+	changeAmtOpt.WhenSome(func(changeAmt btcutil.Amount) {
+		sweepTx.AddTxOut(&wire.TxOut{
+			PkScript: changePkScript,
+			Value:    int64(changeAmt),
+		})
+	})
+
+	// We'll default to using the current block height as locktime, if none
+	// of the inputs commits to a different locktime.
+	sweepTx.LockTime = uint32(t.currentHeight)
+	locktimeOpt.WhenSome(func(locktime int32) {
+		sweepTx.LockTime = uint32(locktime)
+	})
+
+	prevInputFetcher, err := input.MultiPrevOutFetcher(inputs)
+	if err != nil {
+		return nil, 0, fmt.Errorf("error creating prev input fetcher "+
+			"for hash cache: %v", err)
+	}
+	hashCache := txscript.NewTxSigHashes(sweepTx, prevInputFetcher)
+
+	// With all the inputs in place, use each output's unique input script
+	// function to generate the final witness required for spending.
+	addInputScript := func(idx int, tso input.Input) error {
+		inputScript, err := tso.CraftInputScript(
+			t.cfg.Signer, sweepTx, hashCache, prevInputFetcher, idx,
+		)
+		if err != nil {
+			return err
+		}
+
+		sweepTx.TxIn[idx].Witness = inputScript.Witness
+
+		if len(inputScript.SigScript) == 0 {
+			return nil
+		}
+
+		sweepTx.TxIn[idx].SignatureScript = inputScript.SigScript
+
+		return nil
+	}
+
+	for idx, inp := range idxs {
+		if err := addInputScript(idx, inp); err != nil {
+			return nil, 0, err
+		}
+	}
+
+	log.Debugf("Created sweep tx %v for %v inputs", sweepTx.TxHash(),
+		len(inputs))
+
+	return sweepTx, txFee, nil
+}
+
+// prepareSweepTx returns the tx fee, an optional change amount and an optional
+// locktime after a series of validations:
+// 1. check the locktime has been reached.
+// 2. check the locktimes are the same.
+// 3. check the inputs cover the outputs.
+//
+// NOTE: if the change amount is below dust, it will be added to the tx fee.
+func prepareSweepTx(inputs []input.Input, changePkScript []byte,
+	feeRate chainfee.SatPerKWeight, currentHeight int32) (
+	btcutil.Amount, fn.Option[btcutil.Amount], fn.Option[int32], error) {
+
+	noChange := fn.None[btcutil.Amount]()
+	noLocktime := fn.None[int32]()
+
+	// Creating a weight estimator with nil outputs and zero max fee rate.
+	// We don't allow adding customized outputs in the sweeping tx, and the
+	// fee rate is already being managed before we get here.
+	inputs, estimator, err := getWeightEstimate(
+		inputs, nil, feeRate, 0, changePkScript,
```

Comment:
this shadows inputs, according to the operations description it "returns counts for cltv and csv inputs" but that doesn't seem to be what's going on. Why is it returning inputs here rather than just a weight estimate as its name suggests?

Reply:
yeah not sure - also don't like this method, future refactoring PRs tho.